### PR TITLE
add how to create HttpApi plugins for network modules

### PR DESCRIPTION
**Pull request by samccann**
##### SUMMARY

Create a new developer page for how to create HttpAPI plugins for network modules

Partially addresses #53681
##### ISSUE TYPE

- Docs Pull Request

##### COMPONENT NAME

docs.ansible.com
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: docs/docsite/rst/dev_guide/developing_modules_network.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+.. contents::
+   :local:
+
+.. _developing_modules_network:
+
+Network connection plugins
+==========================
+
+Each network connection plugin has a set of implementation plugins which provide a specific implementation of that connection for a particular set of devices.
```
Comment (kbreit) [inline docs/docsite/rst/dev_guide/developing_modules_network.rst]: I don't like 'implementation plugins'. I appreciate that's the phrase, but it's confusing since 'implementation' is used later. Maybe just say "...has a set of plugins which provide a specification of the connection for a particular set of devices."

File: docs/docsite/rst/dev_guide/developing_modules_network.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+--------------------------
+
+:ref:`HttpApi plugins <httpapi_plugins>` serve as adapters for various HTTP(S) APIs for use with the ``httpapi`` connection plugin. They should implement a minimal set of convenience methods tailored to the API you are attempting to use.
+
+Specifically, there are a few methods that the ``httpapi`` connection plugin expects to exist.
+
+Making requests
+^^^^^^^^^^^^^^^
+
+The ``httpapi`` connection plugin has a ``send()`` method, but an HttpApi plugin needs a ``send_request(self, data, **message_kwargs)`` method as a higher-level wrapper to ``send()``. This method should prepare requests by adding fixed values like common headers or URL root paths, and may do more complex work such as turning data into formatted payloads, or determining which path or method to request. It may then also unpack responses to be more easily consumed by the caller.
```
Comment (kbreit) [inline docs/docsite/rst/dev_guide/developing_modules_network.rst]: What's the difference between httpapi and HttpApi?

The second sentence could be cleaned up. Not sure if it's a run-on per ce, but it is too wordy.

Also, this hits one of my pet pieves - using the word 'consume'.

**Comment by kbreit** [comment]
@samccann Overall this looks to be a good start. I left some inline comments but have a few I'm taking out here.

This should have examples in the document. For example, the authentication section should have examples of how these are used. The variables would be dummy variable names, but they'd be obvious to the reader.

Second, I'd like a "Is HttpApi right for you?" section. I keep hearing `connection: local` will be deprecated, but that's probably a while off. This section should have features it supports as well as features it doesn't support.

File: docs/docsite/rst/dev_guide/developing_modules_network.rst (hunk 2 above)
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_network.rst]: Yeah, don't capitalize, there's no value IMO, only confusing.

**Comment by maxamillion** [comment]
👍

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+The ``httpapi`` connection plugin has a ``send()`` method, but an httpapi plugin needs a ``send_request(self, data, **message_kwargs)`` method as a higher-level wrapper to ``send()``. This method should prepare requests by adding fixed values like common headers or URL root paths. This method may do more complex work such as turning data into formatted payloads, or determining which path or method to request. It may then also unpack responses to be more easily consumed by the caller.
+
+.. code-block:: python
+
+   def send_request(self, data, path, method='POST'):
+       # Fixed headers for requests
+       headers = {'Content-Type': 'application/json'}
+       try:
+           response, response_content = self.connection.send(path, data, method=method, headers=headers)
```
Comment (kbreit) [inline docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:57]: This particular example uses a `POST` method so it will contain body, probably represented by data. What happens with a `GET` request?

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+       # handle_response (defined separately) will take the format returned by the device
+       # and transform it into something more suitable for use by modules.
+       # This may be JSON text to Python dictionaries, for example.
+       return handle_response(response_content)
+
+Authenticating
+--------------
+
+By default, all requests will authenticate with HTTP Basic authentication. If a request can return some kind of token to stand in place of HTTP Basic, the ``update_auth(self, response, response_text)`` method should be implemented to inspect responses for such tokens. If the token is meant to be included with the headers of each request, it is sufficient to return a dictionary which will be merged with the computed headers for each request. The default implementation of this method does exactly this for cookies. If the token is used in another way, say in a query string, you should instead save that token to an instance variable, where the ``send_request()`` method (above) can add it to each request
```
Comment (kbreit) [inline docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:69]: Based solely on this paragraph, it's not clear to me how to structure calls which use a static header token and don't need any login or logout calls. The paragraph opens with "By default..." so by just reading this paragraph, I'm concerned I will need to disable that.

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+The ``httpapi`` connection plugin has a ``send()`` method, but an httpapi plugin needs a ``send_request(self, data, **message_kwargs)`` method as a higher-level wrapper to ``send()``. This method should prepare requests by adding fixed values like common headers or URL root paths. This method may do more complex work such as turning data into formatted payloads, or determining which path or method to request. It may then also unpack responses to be more easily consumed by the caller.
+
+.. code-block:: python
+
+   def send_request(self, data, path, method='POST'):
+       # Fixed headers for requests
+       headers = {'Content-Type': 'application/json'}
+       try:
+           response, response_content = self.connection.send(path, data, method=method, headers=headers)
+       except HTTPError as exc:
```
Comment (kbreit) [inline docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:58]: Documentation isn't showing where `HTTPError` is imported from.

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+Each network connection plugin has a set of plugins which provide a specification of the connection for a particular set of devices. Public methods of these plugins may be called on the connection proxy object from the module just as connection methods can.
+
+.. code-block:: python
+
+  from ansible.module_utils.connection import Connection
+
+  ...
+
+  # module is your AnsibleModule instance
+  connection = Connection(module._socket_path)
```
Comment (kbreit) [inline docs/docsite/rst/network/dev_guide/developing_plugins_network.rst]: I am receiving errors around `module._socket_path` not being found (don't have the error in front of me). I'd like to see more documentation about configuring `_socket_path`. I've also noticed not all plugin examples appear to use it (FTD?) so more detail about that may be helpful too.

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst [hunk 7]
```diff
@@ -0,0 +1,108 @@
+
+.. _developing_modules_network:
+.. _developing_plugins_network:
+
+**************************
+Network connection plugins
+**************************
+
+Each network connection plugin has a set of plugins which provide a specification of the connection for a particular set of devices. Public methods of these plugins may be called on the connection proxy object from the module just as connection methods can.
+
```
Comment (kbreit) [inline docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:19]: A paragraph explaining how `Connection` knows about the proper plugin could be helpful to someone newer to the framework. I'll include a proposal.

Comment (kbreit) [inline docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:19]:
```suggestion
Because the httpapi architecture uses plugins, modules do not refer directly to their httpapi implemention for requests. Instead, they reference `Connection`, as shown below. `Connection` uses the host's `ansible_network_os` value to determine which plugin implementation to execute for a host.
```

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
```
Comment (kbreit) [inline docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:1]: Would it make sense to include a reference example implementation that is bare bones and can be used a good start? The examples here are good, but it doesn't demonstrate everything end-to-end.

**Comment by samccann** [comment]
@Qalthos  - can you take a look at the open questions above?

**Comment by kbreit** [comment]
A section explaining how to specify input parameter names (ex. username) would be useful. I am trying to create a key for `api_key` but receiving errors that it's not defined.

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst (hunk 3 above)
Comment (Qalthos) [reply docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:57]: This example defaults to `POST`, but could just as easily be called with `GET` (or `PATCH`, or `PUT`, or `DELETE`), The point is that this method should take whatever information it needs from a module or module_utils and turns that into a valid request for the API it follows. That could involve packing `data` into a JSON payload, or ignoring it if it sees that `method` is `'GET'`, or even just passing it through as shown and trusting whatever calls it to give it the right format. What happens in any case is up to the needs of your plugin, which is why it isn't implemented in the base class.

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst (hunk 4 above)
Comment (Qalthos) [reply docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:69]: I'm not entirely following your question. This is the section that deals with not having login or logout calls. If your header token is already known beforehand and there is no other authentication, then you would have to take some other approach not listed here (probably adding the token as a plugin option, and setting `self.connection._auth` to use the value of that option in `__init__` similar to how the `login()` example does it below)

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst (hunk 3 above)
Comment (kbreit) [reply docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:57]: So this could be a valid implementation? I'm just swagging it, not sure it works...
```
def send_request(self, data, path, method='POST'):
    if method == 'POST':
        response, response_content = self.connection.send(path, data, method=method, headers=headers)
    elif method == 'GET':
        response, response_content = self.connection.send(path, None, method=method, headers=headers)
```

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst (hunk 4 above)
Comment (kbreit) [reply docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:69]: I think that makes sense. I'll need to test my implementation out a little more thoroughly to be sure, but it does mostly answer my question. I wouldn't mind seeing a note stating this as such though.

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst (hunk 3 above)
Comment (Qalthos) [reply docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:57]: I would probably do

```
if method == 'GET':
    data = None
response, response_content = self.connection.send(path, data, method=method, headers=headers)
```

Unless you had some other reason to separate the `send()` calls, but that would work as well.

File: docs/docsite/rst/network/dev_guide/developing_plugins_network.rst (hunk 7 above)
Comment (Qalthos) [reply docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:19]: I've gone a slight;y different direction, first to hopefully make clearer that this section does is about _module_ code, and not plugins, and second because this document is not specifically about `httpapi`, though that is where we are focusing the most attention right now.

Comment (kbreit) [reply docs/docsite/rst/network/dev_guide/developing_plugins_network.rst:19]: I saw the commit you pushed and it was clear. Thank you!

**Review by Qalthos — APPROVED** [review]
